### PR TITLE
FIX #66 - accessing project context

### LINKED ISSUE #66: Accessing the project context inside hooks implementations
Hi, 

I'm a colleague of @Galileo-Galilei. We discussed the issue of accessing the project template and properties from a plugin. [We raised the point in the kedro project](https://github.com/quantumblacklabs/kedro/issues/506)

What will it solve : 

- No more Template assumptions. We will ask the kedro context where the template component resides (project_paths, conf_paths, src path, ...). That lead to less kedro updates regression, because we will no longer manage a template at our side.

- No more configs loading logic assumptions. We will use the kedro context configLoader and get_crendentials (that can be overrided by user).

That solve : #64 #54 #30 #31 (User will use kedro credentials mechanisms, i'll put details in another issue) and a part of #62

How can we implement it (For now, i see three possibilities):

-  Calling a load_context inside before_pipeline_run hook and setting at the same time a global project_path variable, hoping that the before_pipeline_run hook is always the first hook that will be called in the futur. **--> So hacky**

- Declaring the mlflow hooks as a project context property. Which means that the hooks can easily access to the current context **--> No more auto registration**

- Waiting for the release of the kedro session. It will apparently allows access to the current context **--> We don't know when will it land**

Keep up the good work guys, I'm happy to join you :)

### PR DESCRIPTION
**Pull request by takikadiri**
closes #66  #30 #72

## Description
This PR intend to fix mlflow configs related issues.

## Development notes

Some important changes :

* kedro-mlflow can now access to project context properties. That decouple the plugin from the template.
We used this access to change get_mlflow_config signature, the test have been updated accordignately.

* I add a .kedro.yml file, src folder and run.py file to the kedro project fixture for tests running in kedro 0.16.5

For more details, see changes.

Checklist

## Checklist

- [X] Read the [contributing](https://github.com/Galileo-Galilei/kedro-mlflow/blob/develop/CONTRIBUTING.md) guidelines
- [X] Open this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Update the documentation to reflect the code changes
- [X] Add a description of this change and add your name to the list of supporting contributions in the [`CHANGELOG.md`](https://github.com/Galileo-Galilei/kedro-mlflow/blob/develop/CHANGELOG.md) file. Please respect [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines.
- [X] Add tests to cover your changes

## Notice

- [X] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,21 @@
 
 ## [Unreleased]
 
+### Added
+
+-
+
+### Fixed
+
+- `get_mlflow_config` now uses the kedro context config_loader to get configs (#66 #64 #54 #30 #31 #72)
```
Comment (Galileo-Galilei) [inline CHANGELOG.md:11]: Could you change this line to be more explicit about the changes? It ill help users to find out the very important breakthrough you've made, and eventually to look in CHANGELOG later if needed (with a ctrl+f) .Something like (but feel free to change the wording): 

```markdown
- `get_mlflow_config` now uses the kedro context config_loader to get configs (#66). This indirectly solves the following issues:
    -  `get_mlflow_config` now works in interactive mode if `load_context` is called  with a path different from the working directory (#30)
    - kedro_mlflow now works fine with kedro jupyter notebook independently of the working directory (#64) 
    - You can use global variables in `mlflow.yml` which  is now properly parsed if you use a `TemplatedConfigLoader`  (#72)
This change does not fix, but paves the way to further enhancement:
    - Add the possibility to manage environment variables through kedro `credentials.yml` file (#31) 
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,21 @@
...
+
+### Changed
+
+- `MlflowNodeHook` have now a before_pipeline_run hook allowing to load the kedro context, get configs and populating MlflowNodeHook attributes
```
Comment (Galileo-Galilei) [inline CHANGELOG.md:15]: [...] a `before_pipeline_run` hook which stores the  `ProjectContext` and enable to retrieve configuration.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,21 @@
...
+- `mlflow init` is now getting conf path from `context.CONF_ROOT`
```
Comment (Galileo-Galilei) [inline CHANGELOG.md:16]: This goes to "fixed" I think.  

"- `mlflow init` is now getting conf path from `context.CONF_ROOT` instead of hardcoded conf folder. This makes the package robust to kedro changes."

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,6 +2,21 @@
...
+- Change conf_dir fixture in conftest by adding `.kedro.yml`, a `src/dummy_package_run.py` tree structure and `globals.yml`.
+- Change tests that are impacted by the introduction of the kedro context
```
Comment (Galileo-Galilei) [inline CHANGELOG.md:18]: I don't think that tests modifications go the CHANGELOG which is intended for users (but it's nice for the review to tell what you changed ;) I think it rather belongs to the discussion in the PR.

File: kedro_mlflow/framework/hooks/node_hook.py [hunk 5]
```diff
@@ -51,6 +90,9 @@ def before_node_run(
         mlflow.log_params(params_inputs)
 
 
+mlflow_node_hooks = MlflowNodeHook()
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/hooks/node_hook.py:93]: Can you remove this line? It does not belong to this PR, it is linked to the auto-regsitration one.

File: kedro_mlflow/framework/hooks/pipeline_hook.py [hunk 6]
```diff
@@ -12,11 +13,14 @@
 from kedro_mlflow.framework.context import get_mlflow_config
 from kedro_mlflow.io import MlflowMetricsDataSet
 from kedro_mlflow.mlflow import KedroPipelineModel
-from kedro_mlflow.pipeline.pipeline_ml_factory import PipelineML
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/hooks/pipeline_hook.py:15]: Whoops, I wonder why this bug was not spotted by tests or flake8? Well spotted!

File: kedro_mlflow/framework/hooks/pipeline_hook.py [hunk 7]
```diff
@@ -177,6 +185,9 @@ def on_pipeline_error(
             mlflow.end_run()
 
 
+mlflow_pipeline_hooks = MlflowPipelineHook()
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/hooks/pipeline_hook.py:188]: Can you remove this line too? It is also linked to the auto-registration PR.

**Review by Galileo-Galilei — CHANGES_REQUESTED** [review]
Really great! My comments are really minor and just aim at improving readibility Could you please squash your 2 commits? and create the ingle commit "FIX #66 - Accessing project context inside hooks (and consequently fix #30 #64 #72)"

We're almost good to go!

File: CHANGELOG.md (hunk 1 above)
Comment (takikadiri) [reply CHANGELOG.md:11]: yes, it's more informative ! that show concretely what user gain from the changes.
But the last one `- Add the possibility to manage environment variables through kedro credentials.yml file` can be a bit confusing, as it's not fixed yet. Pointing further enhancement enabled (or facilitated) by a change will complexify the changelog readability.
I suggest to adress the last point when it is fixed. probably with resolving this issue #77

File: CHANGELOG.md (hunk 3 above)
Comment (takikadiri) [reply CHANGELOG.md:16]: We didn't raise an issue for that. But yeah, it sound like a fix

File: CHANGELOG.md (hunk 4 above)
Comment (takikadiri) [reply CHANGELOG.md:18]: Ok for me

File: kedro_mlflow/framework/hooks/node_hook.py (hunk 5 above)
Comment (takikadiri) [reply kedro_mlflow/framework/hooks/node_hook.py:93]: It's not just for auto-registration, it's also for declaring hooks in .kedro.yml that we normally support since we moved the hook's configurations into `mlflow.yml`. When the `_register_hooks` is collecting hooks in .kedro.yml or in entry_points, he don't instantiate the hook object, they use our class as is, which raise an error when kedro use the hook, because he don't get the `self` argument.
So it's more a fix for precedent release. We can let this fix for the auto-regsitration PR

File: kedro_mlflow/framework/hooks/pipeline_hook.py (hunk 7 above)
Comment (takikadiri) [reply kedro_mlflow/framework/hooks/pipeline_hook.py:188]: Same for node_hook

File: CHANGELOG.md (hunk 1 above)
Comment (Galileo-Galilei) [reply CHANGELOG.md:11]: Agreed. We'll make a comment in #77 after this PR is merged to keep track that it is now resolvable.

File: CHANGELOG.md (hunk 3 above)
Comment (Galileo-Galilei) [reply CHANGELOG.md:16]: > We didn't raise an issue for that
Because no one has spotted the error before :)

File: kedro_mlflow/framework/hooks/node_hook.py (hunk 5 above)
Comment (Galileo-Galilei) [reply kedro_mlflow/framework/hooks/node_hook.py]: I don't really understand why Kedro would throw an error, and I am not aware of this bug in previous release (I mean that no user gave me such feedback, but I only try it with `kedro>=0.16.0,<0.16.4`). Is it linked to kedro 0.16.5 new hook management? If yes, let's add it to the PR to fix #62 then.

Comment (takikadiri) [reply kedro_mlflow/framework/hooks/node_hook.py]: Oh ! I thought declaring hooks in .kedro.yml was possible on kedro <0.16.5. It's just introduced on 0.16.5
So yeah, we can throw it to #62
